turbopack/crates/turbo-tasks-backend/src/backend/storage.rs

```diff
@@ -0,0 +1,390 @@
+        &mut self,
+        key: &T::Key,
+        update: impl FnOnce(Option<T::Value>) -> Option<T::Value>,
+    ) {
+        let map = self.map_mut(key);
+        if let Some(value) = map.get_mut(key) {
+            let v = take(value);
+            if let Some(v) = update(Some(v)) {
+                *value = v;
+            } else {
+                map.remove(key);
+            }
+        } else if let Some(v) = update(None) {
+            map.insert(key.clone(), v);
+        }
+    }
+}
+
+impl<T: KeyValuePair + Default> InnerStorage<T>
+where
+    T::Key: Indexed,
+    T::Value: PartialEq,
+{
+    pub fn has(&self, item: &mut T) -> bool {
+        let (key, value) = take(item).into_key_and_value();
+        let result = if let Some(stored_value) = self.get(&key) {
+            *stored_value == value
+        } else {
+            false
+        };
+        *item = T::from_key_and_value(key, value);
+        result
+    }
+}
+
+pub struct Storage<K, T: KeyValuePair>
+where
+    T::Key: Indexed,
+{
+    map: DashMap<K, InnerStorage<T>, BuildHasherDefault<FxHasher>>,
+}
+
+impl<K, T> Storage<K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+    K: Eq + std::hash::Hash + Clone,
+{
+    pub fn new() -> Self {
+        let shard_amount =
+            (available_parallelism().map_or(4, |v| v.get()) * 64).next_power_of_two();
+        Self {
+            map: DashMap::with_capacity_and_hasher_and_shard_amount(
+                1024 * 1024,
+                Default::default(),
+                shard_amount,
+            ),
+        }
+    }
+
+    pub fn access_mut(&self, key: K) -> StorageWriteGuard<'_, K, T> {
+        let inner = match self.map.entry(key) {
+            dashmap::mapref::entry::Entry::Occupied(e) => e.into_ref(),
+            dashmap::mapref::entry::Entry::Vacant(e) => e.insert(InnerStorage::new()),
+        };
+        StorageWriteGuard {
+            inner: inner.into(),
+        }
+    }
+
+    pub fn access_pair_mut(
+        &self,
+        key1: K,
+        key2: K,
+    ) -> (StorageWriteGuard<'_, K, T>, StorageWriteGuard<'_, K, T>) {
+        let (a, b) = get_multiple_mut(&self.map, key1, key2, || InnerStorage::new());
+        (
+            StorageWriteGuard { inner: a },
+            StorageWriteGuard { inner: b },
+        )
+    }
+}
+
+pub struct StorageWriteGuard<'a, K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+{
+    inner: RefMut<'a, K, InnerStorage<T>, BuildHasherDefault<FxHasher>>,
+}
+
+impl<'a, K, T> Deref for StorageWriteGuard<'a, K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+    K: Eq + Hash,
+{
+    type Target = InnerStorage<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<'a, K, T> DerefMut for StorageWriteGuard<'a, K, T>
+where
+    T: KeyValuePair,
+    T::Key: Indexed,
+    K: Eq + Hash,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+#[macro_export]
+macro_rules! get {
```

Comment:
Ugh, I hate `#[macro_export]` because it unconditionally makes the macro pub in the root of the crate, but it seems unavoidable here as multiple modules need to access the macro. Rust macros are very half-baked.

- Give this a more descriptive name than just `get!`.
- add `#[doc(hidden)]` since this looks like an internal-only thing.

Same goes for the other macros here.

Reply:
Seems like Rust fixed that problem and one can use `pub(crate) use get;` now instead of macro_export